mule-extensions-api/src/main/java/org/mule/runtime/extension/api/runtime/operation/OperationResult.java

```diff
@@ -0,0 +1,127 @@
+import org.mule.runtime.api.metadata.MediaType;
+
+import java.io.Serializable;
+import java.util.Optional;
+
+/**
+ * Represents the result of an operation execution. Extensions can use this class
+ * for cases in which the operation not only needs to return a value to be set
+ * on the message payload but also wants to specify message attributes and/or
+ * {@link MediaType}.
+ * <p>
+ * The {@link #getOutput()} value is always taken at face value, meaning that
+ * if it's {@code null}, then the message payload will also be nullified. However,
+ * if the {@link #getAttributes()} or {@link #getMediaType()} are {@link Optional#empty},
+ * then the runtime will interpret that as the operation not interested in changing
+ * those values, keeping the input message's attributes and/or media type untouched.
+ *
+ * @param <OUTPUT>     the generic type of the output value
+ * @param <ATTRIBUTES> the generic type of the message attributes
+ * @since 1.0
+ */
+public interface OperationResult<OUTPUT, ATTRIBUTES extends Serializable>
+{
+
+    /**
+     * Creates a new {@link Builder}
+     *
+     * @param <OUTPUT>     the generic type of the output value
+     * @param <ATTRIBUTES> the generic type of the message attributes
+     * @return a new {@link Builder}
+     */
+    static <OUTPUT, ATTRIBUTES extends Serializable> Builder<OUTPUT, ATTRIBUTES> builder()
+    {
+        return OperationResultBuilderFactory.getDefaultFactory().create();
+    }
+
+    /**
+     * Creates a new {@link Builder} initialises with a state that matched
+     * the one of the given {@code muleMessage}
+     *
+     * @param muleMessage  a reference {@link MuleMessage}
+     * @param <OUTPUT>     the generic type of the output value
+     * @param <ATTRIBUTES> the generic type of the message attributes
+     * @return a new {@link Builder}
+     */
+    static <OUTPUT, ATTRIBUTES extends Serializable> Builder<OUTPUT, ATTRIBUTES> builder(MuleMessage muleMessage)
+    {
+        return (Builder<OUTPUT, ATTRIBUTES>) OperationResultBuilderFactory.getDefaultFactory().create()
+                .output(muleMessage.getPayload())
+                .attributes(muleMessage.getAttributes())
+                .mediaType(muleMessage.getDataType().getMediaType());
+    }
+
+    /**
+     * @return The operation's output
+     */
+    OUTPUT getOutput();
+
+    /**
+     * The new value that the operation wants to set on {@link MuleMessage#getAttributes()}.
+     * <p>
+     * The operation might not be interested in changing that value, in which case
+     * this method would return {@link Optional#empty()}
+     *
+     * @return an {@link Optional} {@code ATTRIBUTES} value
+     */
+    Optional<ATTRIBUTES> getAttributes();
+
+    /**
+     * The new {@link MediaType} that the operation wants to set on {@link MuleMessage#getDataType()}.
+     * <p>
+     * The operation might not be interested in changing that value, in which case
+     * this method would return {@link Optional#empty()}
```

Comment:
Going back to the previous comment, I think we need to:
 - If payload doesn't return anything, void, then do not let the guy who code the operation return anything what would means that the message won't be affected at all.
 - if payload was changes by the operation, then we should force the developer to specify a media type.